File: pfb_floating.py
# -*- coding: utf-8 -*-
"""
Created on Thu Aug 16 15:21:43 2018

@author: talonmyburgh
"""
import numpy as np
from pfb_coeff_gen import coeff_gen

# =============================================================================
# Bit reversal algorithms used for the iterative fft's data re-ordering
# =============================================================================
"""Arranges chronological values in an array in a bit reversed fashion"""
def bit_rev(a, bits):
    a_copy = a.copy()
    N = 1<<bits
    for i in range(1,bits):
        a >>=1
        a_copy <<=1
        a_copy |= (a[:]&1)
    a_copy[:] &= N-1
    return a_copy

"""Takes an array of length N which must be a power of two"""
def bitrevarray(array,N): 
    bits = int(np.log2(N))                                                     #number of bits to repr numbers in array
    A = np.empty(N,dtype=np.complex64)
    a=np.arange(N)
    A[bit_rev(a,bits)] = array[:]
    return A

# =============================================================================
# FFT: natural data order in, bit reversed twiddle factors, bit reversed 
# order out.
# =============================================================================
"""Generate array of needed twiddles"""
def make_twiddle(N):                                                           
    i=np.arange(N//2)
    arr = np.exp(-2*i*np.pi*1j/N)
    return arr

"""Natural order in DIT FFT that accepts the data, the twiddle factors
(bit reversed) and allows for staging"""
def iterfft_natural_in_DIT(DATA,twid,staged=False):
    data = np.asarray(DATA,dtype = np.complex64)
    N = data.shape[0]                                                          #how long is data stream
    
    if(staged):
        stgd_data = np.zeros((N,int(np.log2(N))+2),dtype = np.complex64)
        stgd_data[:,0] = data[:]
    num_of_groups = 1                                                          #number of groups - how many subarrays are there?
    distance = N//2                                                            #how far between each fft arm?
    stg=1                                                                      #stage counter
    
    while num_of_groups < N:                                                   #basically iterates through stages
        for k in range(num_of_groups):                                         #iterate through each subarray
            jfirst = 2*k*distance                                              #index to beginning of a group
            jlast = jfirst + distance - 1                                      #first index plus offset - used to index whole group
            W=twid[k]
            slc1 = slice(jfirst,jlast+1,1)
            slc2 = slice(jfirst+distance, jlast+1+distance,1)
            tmp = W*data[slc2]
            data[slc2] = data[slc1]-tmp
            data[slc1] = data[slc1]+tmp
        num_of_groups *=2
        distance //=2
        if(staged):                                                            #if we are recording stages
            stgd_data[:,stg]=data[:]                                           #log each stage data to array
        stg+=1
        
    if(staged): 
        stgd_data[:,-1] = bitrevarray(stgd_data[:,-2],N)                       #post bit-reordering for last stage - added as extra stage
        return stgd_data
    else:
        A=bitrevarray(data,N)                                                  #post bit-reordering
        return A

# =============================================================================
# Floating point PFB implementation making use of the natural order in fft
# like CASPER does. 
# =============================================================================

class FloatPFB(object):
        """This function takes point size, how many taps, what percentage of total data to average over,
        to get data from a file or not,what windowing function, whether you're running dual polarisations,
        whether you'd like data from a stage, and if so which stage - stage 0 being the data in"""
        def __init__(self, N, taps, datasrc = None, w = 'hann',dual = False,
                     staged = False, fwidth=1, chan_acc = False):
            self.N = N                                                         #how many points                                                   #what averaging
            self.dual = dual                                                   #whether you're performing dual polarisations or not
            self.reg =np.zeros([N,taps])                                       #our fir register size filled with zeros orignally
            self.inputdatadir = None
            self.staged=staged
            self.fwidth = fwidth
            self.chan_acc = chan_acc

            if(datasrc is not None and type(datasrc)==str):                    #if input data file is specified
                self.inputdatadir = datasrc
                self.outputdatadir = datasrc[:-4]+"out.npy"
                self.inputdata  = np.load(datasrc, mmap_mode = 'r')
            else:
                self.inputdata = None
            
            self.window = coeff_gen(N,taps,w,self.fwidth)[0]                   #Get window coefficients and scaling 
                                                                               #factor to use in FIR registers.
            self.twids = make_twiddle(self.N)
            self.twids = bitrevarray(self.twids, len(self.twids))              #for natural order in FFT
                
        """Takes data segment (N long) and appends each value to each fir.
        Returns data segment (N long) that is the sum of fircontents*windowcoeffs"""
        def _FIR(self,x):
            self.reg = np.column_stack((x,self.reg))[:,:-1]                    #push and pop from FIR register array
            X = np.sum(self.reg*self.window,axis=1)                            #filter and scale
            return X
        
        """For dual polarisation processing, we need to split the data after
        FFT and return the individual complex spectra"""        
        def _split(self,Y_k):
            R_k = np.real(Y_k)
            R_kflip = R_k.copy()
            R_kflip[1:] = R_kflip[:0:-1]
            
            I_k = np.imag(Y_k)
            I_kflip = I_k.copy()
            I_kflip[1:] = I_kflip[:0:-1]
            
            self.G_k = (1/2)*(R_k+1j*I_k+R_kflip-1j*I_kflip)                   #declares two variables for 2 pols
            self.H_k = (1/2j)*(R_k+1j*I_k-R_kflip+1j*I_kflip)
        
        
        """Here we take the power spectrum of the outputs. Chan_acc dictates
        if one must sum over all outputs produced."""       
        def _pow(self,X):
            if (self.chan_acc):                                                #if accumulation specified
                pwr = X * np.conj(X)
                pwr = np.real(np.sum(pwr,axis=1))
                return pwr
            else:                                                              #if no accumulation specified
                pwr = np.real((X * np.conj(X)))
                return pwr
                        
        """Here one parses a data vector to the PFB to run. Note it must be
        numpy array of length N if a data file was not specified before"""
        def run(self,data=None):
            
            if (data is not None):                                             #if we are using an input data array
                self.inputdata = data
            elif(self.inputdata is None):
                raise ValueError ("No input data for PFB specified.")

            size = self.inputdata.size                                         #get length of data stream
            stages = size//self.N                                              #how many cycles of commutator
            
            if(self.staged):                                                   #if storing staged data
                X = np.empty((self.N,stages,int(np.log2(self.N))+2),
                             dtype = np.complex64)
                                                                               #will be tapsize x datalen/point x stages
                for i in range(0,stages):                                      #for each stage, populate all firs, and run FFT once
                    if(i ==0):
                        X[:,i,:] = iterfft_natural_in_DIT(self._FIR(
                                self.inputdata[0:self.N]),self.twids,
                            self.staged)
                    else:
                        X[:,i,:] = iterfft_natural_in_DIT(self._FIR(
                                self.inputdata[i*self.N:i*self.N+self.N]),self.twids,
                            self.staged)
                
            else:                                                              #if storing staged data
                X = np.empty((self.N,stages),dtype = np.complex64)
                                                                               #will be tapsize x stages
                for i in range(0,stages):                                      #for each stage, populate all firs, and run FFT once
                    if(i == 0):
                        X[:,i] = iterfft_natural_in_DIT(self._FIR(
                                self.inputdata[0:self.N]),
                            self.twids)
                    else:
                        X[:,i] = iterfft_natural_in_DIT(self._FIR(
                                self.inputdata[i*self.N:i*self.N+self.N]),
                            self.twids)
            
            """Decide on how to manipulate and display output data"""
            if(self.dual and not self.staged):                                 #If dual processing but not staged                      
                self._split(X)
                self.G_k_pow = self._pow(self.G_k)
                self.H_k_pow = self._pow(self.H_k)
                
            elif(not self.dual and self.staged):                               #If single pol processing and staged
                self.X_k_stgd = X
                self.X_k_pow = self._pow(X[:,:,-1])
                self.X_k = X[:,:,-1]
                
            elif(self.dual and self.staged):                                   #If dual pol and staged
                self.X_k_stgd = X
                self.split(X[:,:,-1])
                self.G_k_pow = self._pow(self.G_k)
                self.H_k_pow = self._pow(self.H_k)
                
            else:                                                              #If single pol and no staging
                self.X_k = X
                self.X_k_pow = self._pow(X)

File: pfb_floating.jl
import Base: conj
include("pfb_coeff_gen.jl");
"""
```
bitRevArray(array::Array{<:Complex{<:Real}}, N::Integer)::Array{<:Complex{<:Real}}
```
Bit reverses an array's contents according to bitRev of the array's indices.
Takes an array of length N which must be a power of two.
"""
function bitRevArray(array::Array{<:Complex{<:Real}}, N::Integer)::Array{<:Complex{<:Real}}
    return array[bitRev(N)];
end

"""
```
makeTwiddle(N::Integer)::Array{<:Complex{<:Real}}
```
Generates twiddle factors for the natInIterDitFFT FFT.
See also: [`natInIterDitFFT`](@ref)   
"""
function makeTwiddle(N::Integer)::Array{<:Complex{<:Real}}
    i = collect(0:div(N, 2)-1);
    twids = exp.(i .* (-2 * pi * (1im / N)));
    return twids;
end

"""
Floating point PFB implementation making use of the natural order in fft
like CASPER does.
"""
struct FloatPFBScheme
    N           :: Integer;
    dual        :: Bool;
    reg         :: Array{ComplexF64};
    staged      :: Bool;
    fwidth      :: Float64;
    chan_acc    :: Bool;
    window      :: Array{Float64};
    twids       :: Array{ComplexF64};
    function FloatPFBScheme(N::Integer, taps::Integer; w::String="hanning", dual::Bool=false, 
                            staged::Bool=false, fwidth::Float64=1.0, chan_acc::Bool=false)
        new(N,
            dual,
            zeros(ComplexF64,N,taps),
            staged,fwidth,
            chan_acc,
            coeff_gen(N, taps, win=w, fwidth=fwidth)[1],
            bitRevArray(makeTwiddle(N),div(N,2))
        );
    end
end

"""
``` 
natInIterDitFFT(pfbsch::FloatPFBScheme, data::Array{<:Complex{<:Real}}) :: Array{<:Complex}
```
natInIterDitFFT accepts PFB scheme and data.
"""
function natInIterDitFFT(pfbsch::FloatPFBScheme, data::Array{<:Complex}) :: Array{<:Complex}
    data = copy(data);
    N = size(data)[1];
    if pfbsch.staged
        stgd_data = zeros(ComplexF64, N, convert(Int64, log2(N)) + 2);
        stgd_data[:,1] .= data[:];
    end
    num_of_groups = 1;
    distance = div(pfbsch.N,2);
    stg = 2;
    while num_of_groups < pfbsch.N
        for k in 0:num_of_groups-1
            jfirst = (2 * k * distance) + 1;
            jlast = jfirst + distance - 1;
            W = pfbsch.twids[k+1];
            slc1 = jfirst:jlast;
            slc2 = slc1 .+ distance;
            tmp = W .* data[slc2];
            data[slc2] .= data[slc1] .- tmp;
            data[slc1] .= data[slc1] .+ tmp;
        end
        num_of_groups *= 2;
        distance = div(distance, 2);
        if pfbsch.staged
            stgd_data[:,stg] = data[:];
        end
        stg += 1;
    end
    if pfbsch.staged
        stgd_data[:,end] .= bitRevArray(stgd_data[:,end - 1], pfbsch.N);
        return stgd_data;
    else
        return bitRevArray(data, pfbsch.N);
    end
end

"""
``` 
natInIterDitFFT(pfbsch::FloatPFBScheme, data::Array{<:Complex{<:Real}}) :: Array{<:Complex}
```
Overload natInIterDitFFT to accept a PFB scheme and two real data arrays.
"""
function natInIterDitFFT(pfbsch::FloatPFBScheme, data_a::Array{<:Complex{<:Real}}, data_b::Array{<:Complex{<:Real}}) :: Array{<:Complex}
    data = data_a .+ data_b.*im;                                                            #Merge into a complex array for processing.
    data = copy(data);
    N = size(data)[1];
    if pfbsch.staged
        stgd_data = zeros(ComplexF64, N, convert(Int64, log2(N)) + 2);
        stgd_data[:,1] .= data[:];
    end
    num_of_groups = 1;
    distance = div(pfbsch.N,2);
    stg = 2;
    while num_of_groups < pfbsch.N
        for k in 0:num_of_groups-1
            jfirst = (2 * k * distance) + 1;
            jlast = jfirst + distance - 1;
            W = pfbsch.twids[k+1];
            slc1 = jfirst:jlast;
            slc2 = slc1 .+ distance;
            tmp = W .* data[slc2];
            data[slc2] .= data[slc1] .- tmp;
            data[slc1] .= data[slc1] .+ tmp;
        end
        num_of_groups *= 2;
        distance = div(distance, 2);
        if pfbsch.staged
            stgd_data[:,stg] = data[:];
        end
        stg += 1;
    end
    if pfbsch.staged
        stgd_data[:,end] .= bitRevArray(stgd_data[:,end - 1], pfbsch.N);
        # print(stgd_data[:,end])
        return stgd_data;
    else
        return bitRevArray(data, pfbsch.N);
    end
end

"""
```
PFBFir(pfbsch::FloatPFBScheme ,x::Array{<:Complex{<:Real}}) :: Array{<:Complex{<:Real}}
```
FIR: Takes data segment (N long) and appends each value to each fir.
Returns data segment (N long) that is the sum of fircontents*windowcoeffs.
"""
function PFBFir(pfbsch::FloatPFBScheme ,x::Array{<:Complex{<:Real}}) :: Array{<:Complex{<:Real}}
    pfbsch.reg .= hcat(x,pfbsch.reg)[:,1:end-1];
    X = sum(pfbsch.reg .* pfbsch.window,dims=2);
    return X;
end

"""
```
SpecSplit(Y_k::Array{<:Complex})::Tuple{Array{<:Complex},Array{<:Complex}}
```
For dual polarisation processing, we need to split the data after
FFT and return the individual complex spectra.
"""
function SpecSplit(Y_k::Array{<:Complex})::Tuple{Array{<:Complex},Array{<:Complex}}
    R_k = real.(Y_k);
    R_kflip = copy(R_k);
    R_kflip[2:end,:].=R_kflip[end:-1:2,:];

    I_k = imag.(Y_k);
    I_kflip = copy(I_k);
    I_kflip[2:end,:].=I_kflip[end:-1:2,:];
    
    G_k = (1/2) .* (R_k .+ im .* I_k .+ R_kflip .- im .* I_kflip);
    H_k = (1/2im) .* (R_k .+ im .* I_k .- R_kflip .+ im .* I_kflip);
    return (G_k, H_k);
end

"""
```
SpecPow(pfbsch::FloatPFBScheme, X::Array{<:Complex}) :: Array{<:Real}
```
Here we take the power spectrum of the outputs. Chan_acc dictates
if one must sum over all outputs produced.
"""
function SpecPow(pfbsch::FloatPFBScheme, X::Array{<:Complex}) :: Array{<:Real}
    if pfbsch.chan_acc
        pwr = X .* conj.(X);
        pwr = real.(sum(pwr,dims=2));
        return pwr;
    else
        pwr = real.(X .* conj.(X));
        return pwr;
    end; 
end;

Base.length(pfbsch::FloatPFBScheme) = 1;
Base.iterate(pfbsch::FloatPFBScheme) = (pfbsch, nothing);
Base.iterate(pfbsch::FloatPFBScheme, state::Nothing) = nothing;

"""
```
RunPFB(pfbsch::FloatPFBScheme; data::Array{<:Complex}=Nothing) 
    :: Union{Tuple{Array{<:Real},Array{<:Real}},Array{<:Real}}
```
Here one parses a data vector to the PFB to run. Note its length must be a multiple
of N if a data file was not specified before in the scheme.
"""
function RunPFB(pfbsch::FloatPFBScheme; data::Array{<:Complex}=Nothing) #:: Union{Tuple{Array{<:Real},Array{<:Real}},Array{<:Real}}
    if isnothing(data)
        #Here we would look to the input file TODO.
    else                                                        #if we are using an input data array
        inputdata = data;
    end;
    size = length(data);                                    #get length of data stream
    com_cycles = div(size,pfbsch.N);                        #how many cycles of commutator
    data_slc = (1-pfbsch.N):0;                              #start with negative slice to make for loops simpler
    if pfbsch.staged                                            #if storing staged data
        X = Array{ComplexF64}(undef,(pfbsch.N,com_cycles,Int(log2(pfbsch.N)+2)));
                                                                #will be tapsize x datalen/point x stages
        for i in 1:com_cycles                                   #for each stage, populate all firs, and run FFT once
            X[:,i,:] .= natInIterDitFFT(pfbsch,PFBFir(pfbsch,inputdata[data_slc.+(pfbsch.N*(i))]));
        end;
    else                                                        #if storing staged data
        X = Array{ComplexF64}(undef,(pfbsch.N,com_cycles));     #will be tapsize x stages
        for j in 1:com_cycles                                   #for each stage, populate all firs, and run FFT once
            X[:,j] .= natInIterDitFFT(pfbsch,PFBFir(pfbsch,inputdata[data_slc.+(pfbsch.N*(j))]));
        end;
    end;
    if pfbsch.dual & ~pfbsch.staged                             #If dual processing but not staged                      
        # return Tuple(SpecPow.(pfbsch, SpecSplit(X)));
        return SpecSplit(X);
    elseif ~pfbsch.dual & pfbsch.staged                         #If single pol processing and staged
        return SpecPow(pfbsch, X[:,:,end]);
    elseif pfbsch.dual & pfbsch.staged                          #If dual pol and staged                
        return Tuple(SpecPow.(pfbsch, SpecSplit(X[:,:,end])));
    else                                                        #If single pol and no staging
        return SpecPow(pfbsch, X);
    end;
end;